File: LeetCode/src/designclass/MinStack.java
package designclass;

import java.util.Stack;

public class MinStack {

    private Stack<Integer> stack;
    private Stack<Integer> minStack;

    public MinStack() {
        stack = new Stack<>();
        minStack = new Stack<>();
    }

    public void push(int val) {
        stack.push(val);
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    public void pop() {
        if (stack.pop().equals(minStack.peek())) {
            minStack.pop();
        }
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }

    public static void main(String[] args) {
        MinStack minStack = new MinStack();

        // Test case
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-3);

        System.out.println("Minimum Element in the Stack: " + minStack.getMin()); // Output: -3

        minStack.pop();

        System.out.println("Top Element in the Stack: " + minStack.top()); // Output: 0
        System.out.println("Minimum Element in the Stack: " + minStack.getMin()); // Output: -2
    }
}

File: LeetCode_Python/designclass/MinStack.py
class MinStack:
    def __init__(self):
        self.min_stack = []
        self.stack = []
        
    def push(self, val: int) -> None:
        self.stack.append(val)
        if not self.min_stack or val <= self.min_stack[-1]:
            self.min_stack.append(val)

    def pop(self) -> None:
        if self.min_stack[-1] == self.stack[-1]:
            self.min_stack.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

class MinStackByTuple:
    def __init__(self):
        self.stack = []
        
    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))
        else:
            self.stack.append((val, min(val, self.stack[-1][1])))

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]

if __name__ == '__main__':
    minStack = MinStack()
    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    print("getMin() -> Expected: -3, Actual:", minStack.getMin())  
    minStack.pop()
    print("top() -> Expected: 0, Actual:", minStack.top())         
    print("getMin() -> Expected: -2, Actual:", minStack.getMin())   

    minStack = MinStackByTuple()
    minStack.push(-2)
    minStack.push(0)
    minStack.push(-3)
    print("getMin() -> Expected: -3, Actual:", minStack.getMin())
    minStack.pop()
    print("top() -> Expected: 0, Actual:", minStack.top())
    print("getMin() -> Expected: -2, Actual:", minStack.getMin()) 
